anyway/flask_app.py

```diff
@@ -1648,3 +1627,118 @@ def embedded_reports_api():
+        )
+    return jsonify({"have_email": bool(get_current_user_email())})
+
+
+# This code is also used as part of the user first registration
+@app.route("/user/update", methods=["POST"])
+def user_update() -> Response:
+    allowed_fields = [
+        "first_name",
+        "last_name",
+        "email",
+        "phone",
+        "user_type",
+        "user_work_place",
+        "user_url",
+        "user_desc",
+    ]
+
+    if current_user.is_anonymous:
+        return Response(
+            "User not logged in.",
+            status=401,
+        )
+
+    reg_dict = request.json
+    if not reg_dict:
+        return Response(
+            "Bad Request (not a JSON or mimetype does not indicate JSON).",
+            status=400,
+        )
+
+    for key in reg_dict:
+        if key not in allowed_fields:
+            return Response(
+                f"Bad Request (Unknown field {key}).",
+                status=400,
+            )
+
+    first_name = reg_dict.get("first_name")
+    last_name = reg_dict.get("last_name")
+    if not first_name or not last_name:
+        return Response(
+            "Bad Request (first name or last name is missing).",
+            status=400,
+        )
+
+    # If we don't have the user email then we have to get it else only update if the user want.
+    tmp_given_user_email = reg_dict.get("email")
+    user_db_email = get_current_user_email()
+    if not user_db_email or tmp_given_user_email:
+        if not tmp_given_user_email:
+            return Response(
+                "Bad Request (There is no email in our DB and there is no email in the json).",
+                status=400,
+            )
+
+        is_valid = validate_email(
+            email_address=tmp_given_user_email,
+            check_regex=True,
+            check_mx=False,
+            use_blacklist=False,
+        )
+        if not is_valid:
+            return Response(
+                "Bad Request (Bad email address).",
+                status=400,
+            )
+
+        user_db_email = tmp_given_user_email
+
+    phone = reg_dict.get("phone")
+    if phone:
+        is_valid = is_valid_number(phone)
+
+        if not is_valid:
+            return Response(
+                "Bad Request (Bad phone number).",
+                status=400,
+            )
+
+    user_type = reg_dict.get("user_type")
+    user_work_place = reg_dict.get("user_work_place")
+    user_url = reg_dict.get("user_url")
+    user_desc = reg_dict.get("user_desc")
+
+    current_user.first_name = first_name
+    current_user.last_name = last_name
+    current_user.email = user_db_email
+    current_user.phone = phone
+    current_user.user_type = user_type
+    current_user.work_on_behalf_of_organization = user_work_place
+    current_user.user_url = user_url
+    current_user.user_desc = user_desc
+    current_user.is_user_completed_registration = True
+    session_commit()
+
+    return Response(status=200)
+
+
+def session_commit() -> None:
```

Comment:
1 - Currently there is only one place that i am testing code that writes to the DB, in the future more code that will be tested and so the use of this function will increase.
 2 - I am not familiar with this concept, can you please explain more? what are the reasons behind it?
*The main reason that i created this function is for mocking(as we talked previously).

Reply:
The reason to have close control over start and end of transaction is that during a transaction
1 - you do not know what was already written, so if some code that is not aware of the transaction is reading, it is not possible to know whether it reads before or after the write. And
2 - parts of the fields / tables may be locked, and this can interfere with other processes that tries to use these tables / fields.

Regarding mocking, I apologize for not being clear.
A technique you can use is that the method that accessed the DB receive a `session=None` as a parameter. If it it empty, it it initializes it in the usual way (e.g. `session = db.session`). In test, the test code generates a session (mock, or session to a dummy DB for testing), and passes it. A fuller explanation in [stackExchange](https://stackoverflow.com/questions/57193597/mocking-a-sqlalchemy-session-for-pytest)

Reply:
About mocking technique - its less relevant in my test case as i simulate browser access - so my params are only limited to what i can pass in the HTTP GET.
But i think i get the idea, Thanks!